jane/doc/extensions/kinds.md

```diff
@@ -0,0 +1,335 @@
+in structures and signatures.  For example, we could have this alias:
+```ocaml
+kind_abbrev_ portable_value_pair = (value & value) mod portable
+```
+
+# Inclusion and variance
+
+This is accepted:
+
+```ocaml
+module M1 : sig
+  type t : value
+end = struct
+  type t = int
+end
+```
+
+This makes sense because the kind of `int` is `immediate`, which is a subkind
+of `value`. Even though users of `M1.t` will be expecting a `value`, the `immediate`
+they get works great. Thus, the kinds of type declarations are *covariant* in the module
+inclusion check: a module type `S1` is included in `S2` when the kind of a type `t`
+in `S1` is included in the kind of `t` in `S2`.
+
+Similarly, this is accepted:
+
+```ocaml
+module M2 : sig
+  type ('a : immediate) t
+end = struct
+  type ('a : value) t
+end
+```
+
+This makes sense because users of `M2.t` are required to supply an `immediate`; even
+though the definition of `M2.t` expects a `value`, the `immediate` it gets works great.
+Thus, the kinds of type declaration arguments are *contravariant* in the module
+inclusion check: a module type `S1` is included in `S2` when the kind of the argument
+to a type `t` in `S2` is included in the kind of that argument in `S1`.
+
+Contravariance in type arguments allows us to have
+
+```ocaml
+module Array : sig
+  type ('a : any) t = 'a array
+  (* ... *)
+end
+```
+
+and still pass `Array` to functors expecting a `type 'a t`, which assumes `('a :
+value)`.
+
+CR ccasinghino: I just stole the above from the layouts docs, verbatim, except
+changing the word layout to kind.  Probably we should delete it from there and
+point here?
+
+Relatedly, a `with type` constraint on a signature can fill in a type with one
+that has a more specific kind.  E.g., this is legal:
+```ocaml
+module type S_any = sig
+  type t : any
+end
+
+module type S_imm = S_any with type t = int
+```
+
+This can be particularly useful for common signatures that might be implemented
+by types with any kind (e.g., `Sexpable`).
+
+# With kinds
+
+Sometimes the kind of a type constructor depends on the kinds of the types that
+instantiate its parameters.  For example, the type `'a list` can mode cross on
+the portability axis if `'a` does.
+
+We could have a `list` whose kind is restricted to work on types that more cross
+on the portability axis, to record this fact, as in:
+```ocaml
+type ('a : value mod portable) portable_list : value mod portable
+```
+But it would be annoying to have many different list types, and we certainly
+don't want to restrict the normal `list` type to work only on a subset of
+`value`s.
+
+The solution to this problem is "with kinds": kinds that record dependencies on
+types.  The actual kind of list in our.
+```ocaml
+type 'a list : value mod non_null (contended portable many with 'a)
+```
+
+CR ccasinghino: I'm giving up on explaining with-kinds here for now. I think
+there is no way to write the kind of list in the current system? (The above is
+not legal syntax)
```

Comment:
It's just `value mod contended portable many with 'a`. That works actually for two different reasons:

1. `value` already has `mod non_null` in it.
2. `with`-bounds don't affect nullability, by special decree. (This is in fact implemented.)

One possible way forward is to move nullability into the layout. But I think we probably won't do that in time for release.